room/runtime/src/main/java/androidx/room/SQLiteCopyOpenHelper.java

```diff
@@ -211,10 +207,66 @@ private void copyDatabaseFile(File destinationFile) throws IOException {
                     + destinationFile.getAbsolutePath());
         }
 
+        // Temporarily open intermediate file database using FrameworkSQLiteOpenHelper and call
+        // open pre-packaged callback. If it fails then intermediate file won't be copied making
+        // invoking pre-packaged callback a transactional operation.
+        callOnOpenPrepackagedDatabaseCallbacks(intermediateFile, writable);
+
         if (!intermediateFile.renameTo(destinationFile)) {
             throw new IOException("Failed to move intermediate file ("
                     + intermediateFile.getAbsolutePath() + ") to destination ("
                     + destinationFile.getAbsolutePath() + ").");
         }
     }
+
+    private Integer readDatabaseVersion(File databaseFile) {
+        try {
+            return DBUtil.readVersion(databaseFile);
+        } catch (IOException e) {
+            Log.w(Room.LOG_TAG, "Unable to read database version.", e);
+            return null;
+        }
+    }
+
+    private void callOnOpenPrepackagedDatabaseCallbacks(File databaseFile, boolean writable) {
+        if(mDatabaseConfiguration.prepackagedCallback != null) {
+
+            SupportSQLiteOpenHelper helper = createFrameworkOpenHelper(databaseFile);
+            try {
+                SupportSQLiteDatabase db = writable ? helper.getWritableDatabase() :
+                        helper.getReadableDatabase();
+
+                mDatabaseConfiguration.prepackagedCallback.onOpenPrepackagedDatabase(db);
+            } catch (SQLiteException e) {
+                throw new RuntimeException("Unable to invoke pre-packaged callback.", e);
+            } finally {
+                // Close the db and let Room re-open it through a normal path
+                helper.close();
+            }
+        }
+    }
+
+    private SupportSQLiteOpenHelper createFrameworkOpenHelper(File databaseFile) {
+        String databaseName = databaseFile.getName();
+        Integer version = readDatabaseVersion(databaseFile);
+        if(version == null) {
+            throw new RuntimeException("Unable to invoke pre-packaged callback.");
```

Comment:
Let's change the message to 'Malformed database file, unable to read version'. It would also be better to use the `DBUtil` method directly here and catch the IOException so that you can propagate it with the new RuntimeException, it provides a bit more info if something goes wrong.